x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportInternalInferModelAction.java

```diff
@@ -348,15 +349,24 @@ public void onFailure(Exception e) {
             }
 
             private void sendResponse() {
-                if (results.nonNullLength() > 0) {
+                if (failure.get() != null) {
+                    finalListener.onFailure(failure.get());
+                } else {
                     for (int i = 0; i < results.length(); i++) {
-                        if (results.get(i) != null) {
-                            responseBuilder.addInferenceResults(results.get(i));
+                        var resultList = results.get(i);
+                        if (resultList != null) {
+                            for (var result : resultList) {
+                                if (result instanceof ErrorInferenceResults errorResult) {
+                                    // Any failure fails all requests
+                                    // TODO is this the correct behaviour for batched requests?
+                                    finalListener.onFailure(errorResult.getException());
```

Comment:
I don't know the code well enough but maybe in the future we could make the response similar to a bulk response where an entry in the results array can either be a failure or a successful result?

Reply:
That the idea. The rest response does not have to change but internal users (such as ingest) can make better decisions about how to handle a response which is partially successful